docs/core/v3.0.rst

```diff
@@ -689,6 +687,12 @@ arbitrary length in a "negative" direction along any dimension.
    in which case the coordinate of a chunk is the empty tuple, and the chunk key
    will consist of the string ``c``.
 
+.. note:: Chunks at the border of an array always have the full chunk size, even when
+   the array only covers parts of it. For example, having and array with ``shape=(30, 30)``,
+   ``chunks=(16, 16)``, the chunk ``0,1`` would also contain unused values for the indices
+   ``(0-16, 30-31)``. Since reading anything beyond the array shape is undefined behavior,
+   those values may be passed on, be replaced with the fill values or other undefined behavior.
```

Comment:
Also this example is using zarr v2 metadata and Python syntax.  It should instead use zarr v3 metadata and json syntax.

As far as "reading anything beyond the array shape is undefined behavior" --- that seems like an implementation concern more than a spec concern.  E.g. some implementations might return an error, while others might do something else.

But the content of those "unused values" stored in the chunk is not necessarily unimportant --- if you resize an array to have larger bounds, and (as is likely to be the case) do not rewrite existing chunks, then what were previously unused values become valid values.  I think, therefore, that the most natural choice for those "unused values" in the chunk is that they are set to the fill value.

Reply:
> Also this example is using zarr v2 metadata and Python syntax. It should instead use zarr v3 metadata and json syntax.

Good point :+1:

> As far as "reading anything beyond the array shape is undefined behavior" --- that seems like an implementation concern more than a spec concern. E.g. some implementations might return an error, while others might do something else.

I think that's exactly what undefined behavior means, that implementations are free to do anything. It's only undefined regarding the spec, and implementations might choose to define it as they want.

> But the content of those "unused values" stored in the chunk is not necessarily unimportant --- if you resize an array to have larger bounds, and (as is likely to be the case) do not rewrite existing chunks, then what were previously unused values become valid values. I think, therefore, that the most natural choice for those "unused values" in the chunk is that they are set to the fill value.

I would rather not put that as a requirement, since it would also require implementations to re-write chunks when resizing an array to smaller bounds, which is not strictly needed and might not be wanted for performance reasons or to be able to restore the previous content when resizing to the previous bound again. Since there are multiple possible valid strategies I'd rather not specify an explicit one in the spec.

Reply:
Agreed that we don't want to require implementations to rewrite chunks when shrinking the bounds --- indeed TensorStore does not rewrite chunks when shrinking the bounds.  When writing initially, though, filling out-of-bounds locations with the fill value seems like an obvious choice, without any real downside.  Because the implementation that initially writes the chunk may not be the same as the implementation that does the resize, I think it is still useful to specify the behavior.

Reply:
Added this as a recommendation in 4389e12.